enterprise_gateway/services/kernelspecs/kernelspec_cache.py

```diff
@@ -0,0 +1,282 @@
+                cache_item = KernelSpecCache.kernel_spec_to_cache_item(cache_item)
+
+            resource_dir = cache_item['resource_dir']
+            self.cache_items[kernel_name.lower()] = cache_item
+            observed_dir = os.path.dirname(resource_dir)
+            if observed_dir not in self.observed_dirs:
+                # New directory to watch, schedule it...
+                self.log.debug("KernelSpecCache: observing directory: {observed_dir}".format(observed_dir=observed_dir))
+                self.observed_dirs.add(observed_dir)
+                self.observer.schedule(KernelSpecChangeHandler(self), observed_dir, recursive=True)
+
+    def put_all_items(self, kernelspecs: Dict[str, CacheItemType]) -> None:
+        """Adds or updates a dictionary of kernel specification in the cache. """
+        if self.cache_enabled and kernelspecs:
+            for kernel_name, cache_item in kernelspecs.items():
+                self.put_item(kernel_name, cache_item)
+
+    def remove_item(self, kernel_name: str) -> Optional[CacheItemType]:
+        """Removes the cache item corresponding to kernel_name from the cache."""
+        cache_item = None
+        if self.cache_enabled:
+            if kernel_name.lower() in self.cache_items:
+                cache_item = self.cache_items.pop(kernel_name.lower())
+                self.log.info("KernelSpecCache: removed kernelspec: {kernel_name}".format(kernel_name=kernel_name))
+        return cache_item
+
+    def _initialize(self):
+        """Initializes the cache and starts the observer. """
+
+        # The kernelspec cache consists of a dictionary mapping the kernel name to the actual
+        # kernelspec data (CacheItemType).
+        self.cache_items = {}  # Maps kernel name to kernelspec
+        self.observed_dirs = set()  # Tracks which directories are being watched
+        self.cache_misses = 0
+
+        # Seed the cache and start the observer
+        if self.cache_enabled:
+            self.observer = Observer()
+            kernelspecs = self.kernel_spec_manager.get_all_specs()
+            self.put_all_items(kernelspecs)
+            # Following adds, see if any of the manager's kernel dirs are not observed and add them
+            for kernel_dir in self.kernel_spec_manager.kernel_dirs:
+                if kernel_dir not in self.observed_dirs:
+                    if os.path.exists(kernel_dir):
+                        self.log.info("KernelSpecCache: observing directory: {kernel_dir}".
+                                      format(kernel_dir=kernel_dir))
+                        self.observed_dirs.add(kernel_dir)
+                        self.observer.schedule(KernelSpecChangeHandler(self), kernel_dir, recursive=True)
+                    else:
+                        self.log.warn("KernelSpecCache: kernel_dir '{kernel_dir}' does not exist"
+                                      " and will not be observed.".format(kernel_dir=kernel_dir))
+            self.observer.start()
```

Comment:
What happens if I go to one of the `jupyter --paths` directory that can be home to a kernelspec and add a new one there (e.g. all kernels were in `/Users/lresende/opt/anaconda/envs/dev/share/jupyter/kernels/` and I add a new one to `/usr/local/share/jupyter/kernels/`? Will the cache/observer identify the new kernelspec?

Reply:
The `self.kernel_spec_manager.kernel_dirs` contains the list of directories in which kernelspecs can reside.  The rub comes if any of those directories do not exist - since a non-existent directory cannot be observed.  However, any kernelspecs added to existing (observed) directories will be detected and automatically added to the cache.

If one of the `kernel_dirs` directories gets created and a kernelspec is added, the kernelspec will not be added to the cache until a cache-miss occurs on that named kernelspec.  However, the problem there is that it doesn't get added to the list of kernelspecs until its found, so end users won't know the new kernelspec is present.  The cache can be 'tickled' by triggering a cache miss using `/api/kernelspecs/new-spec-name` but that probably won't happen.

Of the 4 or 5 default runtime directories, I find only `/usr/share/jupyter/kernels` not existing.  However the more common `/usr/local/share/jupyter/kernels` directory is probably where those kernelspecs are persisted.

We could have another 'monitor' for any `kernel_dirs` that do not exist and add them to the observed list.  This check could be made whenever `get_all_specs` is called - or perhaps use a periodic poll for that.  I suppose if EG is started prior to any kernelspecs being created, this could be a problem, so it's may be worthwhile handling this one scenario.